File: python/project-euler/PE039.py
# -*- coding: utf-8 -*-
"""
If p is the perimeter of a right angle triangle with integral
length sides, {a,b,c}, there are exactly three solutions for p = 120.

{20,48,52}, {24,45,51}, {30,40,50}

For which value of p ≤ 1000, is the number of solutions maximised?
"""


import math


class PE039:
  def triangles(self, p):
    solutions = []
    for a in range(1, int(p / (2 + math.sqrt(2)))):
      for b in range(a, int((p - a) / 2) + 1):
        c = p - a - b
        if c == math.sqrt(a * a + b * b):
          solutions.append([a, b, c])
    return solutions

  def maxTriangles(self):
    max = 0
    for p in range(12, 1001):
      current = len(self.triangles(p))
      if current > max:
        max = current
        maxp = p
    return [max, maxp]

  @staticmethod
  def main():
    test = PE039()
    print(test.triangles(120))  # [[20,48,52], [24,45,51], [30,40,50]]
    print(test.maxTriangles())  # [8, 840]


# test
PE039.main()

File: javascript/project-euler/PE039.js
/*
If p is the perimeter of a right angle triangle with integral 
length sides, {a,b,c}, there are exactly three solutions for p = 120.

{20,48,52}, {24,45,51}, {30,40,50}

For which value of p ≤ 1000, is the number of solutions maximised?
*/

class PE039 {
  triangles(p) {
    const solutions = [];
    for (let a = 1; a <= p / (2 + Math.sqrt(2)); a++) {
      for (let b = a; b < (p - a) / 2; b++) {
        let c = p - a - b;
        if (c === Math.sqrt(a * a + b * b)) {
          solutions.push([a, b, c]);
        }
      }
    }
    return solutions;
  }

  maxTriangles() {
    let max = 0, maxp;
    for (let p = 12; p <= 1000; p++) {
      let current = this.triangles(p).length;
      if (current > max) {
        max = current;
        maxp = p;
      }
    }
    return maxp;
  }

  static main() {
    const test = new PE039();
    console.log(120 / (2 + Math.sqrt(2)))
    console.log(test.triangles(120)); // 
    console.log(test.maxTriangles()); //
    console.log(test.triangles(840)); //
  }
}


// test
PE039.main();